flink-connectors/flink-connector-rabbitmq2/src/main/java/org/apache/flink/connector/rabbitmq2/sink/writer/specialized/RabbitMQSinkWriterExactlyOnce.java

```diff
@@ -0,0 +1,139 @@
+
+    // All messages that arrived and could not be committed this far.
+    private List<RabbitMQSinkMessageWrapper<T>> messages;
+
+    /**
+     * Create a new RabbitMQSinkWriterExactlyOnce.
+     *
+     * @param connectionConfig configuration parameters used to connect to RabbitMQ
+     * @param queueName name of the queue to publish to
+     * @param serializationSchema serialization schema to turn elements into byte representation
+     * @param publishOptions optionally used to compute routing/exchange for messages
+     * @param returnListener return listener
+     */
+    public RabbitMQSinkWriterExactlyOnce(
+            RabbitMQConnectionConfig connectionConfig,
+            String queueName,
+            SerializationSchema<T> serializationSchema,
+            RabbitMQSinkPublishOptions<T> publishOptions,
+            SerializableReturnListener returnListener) {
+        super(connectionConfig, queueName, serializationSchema, publishOptions, returnListener);
+        messages = Collections.synchronizedList(new ArrayList<>());
+    }
+
+    /**
+     * On recover the messages are set to the outstanding messages from the states.
+     *
+     * @param states a list of states to recover the reader with
+     */
+    @Override
+    public void recoverFromStates(List<RabbitMQSinkWriterState<T>> states) {
+        List<RabbitMQSinkMessageWrapper<T>> messages =
+                Collections.synchronizedList(new ArrayList<>());
+        for (RabbitMQSinkWriterState<T> state : states) {
+            messages.addAll(state.getOutstandingMessages());
+        }
+        this.messages = messages;
+    }
+
+    @Override
+    protected void configureChannel() throws IOException {
+        // puts channel in commit mode
+        getRmqChannel().txSelect();
+    }
+
+    @Override
+    public void write(T element, Context context) {
+        messages.add(
+                new RabbitMQSinkMessageWrapper<>(
+                        element, getSerializationSchema().serialize(element)));
+    }
+
+    @Override
+    public List<RabbitMQSinkWriterState<T>> snapshotState() {
+        commitMessages();
+        return Collections.singletonList(new RabbitMQSinkWriterState<>(messages));
+    }
+
+    private void commitMessages() {
+        List<RabbitMQSinkMessageWrapper<T>> messagesToSend = messages.subList(0, messages.size());
+        try {
+            for (RabbitMQSinkMessageWrapper<T> msg : messagesToSend) {
+                getRmqSinkConnection().send(msg);
+            }
+            getRmqChannel().txCommit();
+            LOG.info("Successfully committed {} messages.", messagesToSend.size());
+            messagesToSend.clear();
+        } catch (IOException e) {
+            LOG.error(
+                    "Error during commit of {} messages. Rollback Messages. Error: {}",
+                    messagesToSend.size(),
+                    e.getMessage());
+            try {
+                getRmqChannel().txRollback();
```

Comment:
We added a Runtime exception, so as soon as a transactional commit is not working we will throw it immediately. I guess the is the same discussion as the maxRetry counter we had earlier for sending an at-least-once message. You probably do not want to fail immediately but rather at some point in the future. Like you said this could cause a heavily increased checkpoint size.